Create page for CSPViolationReportBody API (#23433)
mdn/content: files/en-us/web/api/cspviolationreportbody/index.md

- {{DOMxRef("CSPViolationReportBody.toJSON()")}} {{Experimental_Inline}}
  - : A _serializer_ which returns a JSON representation of the `CSPViolationReportBody` object.

## Examples

### Obtaining a `CSPViolationReportBody` object

To obtain a `CSPViolationReportBody` object, you must configure your page so that a CSP violation will occur. In this example, we will set our CSP to only allow content from the site's own origin, and then attempt to load a script from `apis.google.com`, which is an external origin.

First, we will set our {{HTTPHeader("Content-Security-Policy")}} header:

```http
Content-Security-Policy: default-src 'self';
```

Then, we will attempt to load an external script:

```html
<!-- This should generate a CSP violation -->
<script src="https://apis.google.com/js/platform.js"></script>
```

Finally, we will create a new {{domxref("ReportingObserver")}} object to listen for CSP violations.

```js
const observer = new ReportingObserver(
  (reports, observer) => {
    const cspViolation = reports[0].body;
  },
  {
    types: ["csp-violation"],
    buffered: true,
  }
);

observer.observe();
```

> 

Reviewer: Would it be helpful to include a sample JSON payload here?

```js
{
  "csp-report": {
    "blocked-uri": "https://apis.google.com/js/platform.js",
    "disposition": "enforce",
    "document-uri": "<uri-with-csp>",
    "effective-directive": "script-src-elem",
    "original-policy": "default-src 'self'; report-uri <csp-report-uri>",
    "referrer": "",
    "status-code": 200,
    "violated-directive": "script-src-elem"
  }
}
```

Author: I don't think we should include an example of the JSON payload; up at the top, there's a note that indicates this interface is "similar, but not identical to [the JSON payload]", and I feel that including an example of the payload when we're talking about the interface would be confusing.  We also have a sample payload available on the related page: https://developer.mozilla.org/en-US/docs/Web/HTTP/CSP#sample_violation_report

---
title: CSPViolationReportBody
slug: Web/API/CSPViolationReportBody
page-type: web-api-interface
tags:
  - Experimental
browser-compat: api.CSPViolationReportBody
---

{{MDNSidebar}}

> {{APIRef("Reporting API")}}{{SeeCompatTable}}{{SecureContext_Header}}

Author: Done!

> {{APIRef("Reporting API")}}{{SeeCompatTable}}{{SecureContext_Header}}

Reviewer 2: The interface still needs to be listed in https://developer.mozilla.org/en-US/docs/Web/API/Reporting_API.

> {{APIRef("Reporting API")}}{{SeeCompatTable}}{{SecureContext_Header}}

Author: I had just added it to that page in the latest commit!

> {{APIRef("Reporting API")}}{{SeeCompatTable}}{{SecureContext_Header}}

Reviewer 2: It's [conventional in our API overview pages to have a section "Interfaces" that lists all the interfaces in the API](https://developer.mozilla.org/en-US/docs/MDN/Writing_guidelines/Page_structures/Page_types/API_landing_page_template#interfaces). See, for example, https://developer.mozilla.org/en-US/docs/Web/API/Geolocation_API#interfaces.

This list should match the list in the sidebar (which is pulled from GroupData).

In this page it's called "[Reporting API interfaces](https://pr23433.content.dev.mdn.mozit.cloud/en-US/docs/Web/API/Reporting_API#reporting_api_interfaces)" instead, for no good reason, but if we want to treat `CSPViolationReportBody` as an interface of the Reporting API we should list it there (we should also rename it to just "Interfaces" and list the other bodies there, but that's technically out of scope for this PR). We should probably have that instead of https://pr23433.content.dev.mdn.mozit.cloud/en-US/docs/Web/API/Reporting_API#available_report_types.

> {{APIRef("Reporting API")}}{{SeeCompatTable}}{{SecureContext_Header}}

Author: The "Available report types" is a subsection of the "Reporting API interfaces" section, though.  I feel that if we want to make any changes to the structure, we should do so in a separate PR since this PR pertains only to adding `CSPViolationReportBody`.

> {{APIRef("Reporting API")}}{{SeeCompatTable}}{{SecureContext_Header}}

Reviewer 2: OK.

> {{APIRef("Reporting API")}}{{SeeCompatTable}}{{SecureContext_Header}}

Reviewer 2: I'll file a follow-up.

  - : The column number in the document or worker at which the violation occurred.
- {{domxref("CSPViolationReportBody.disposition")}} {{ReadOnlyInline}} {{Experimental_Inline}}
  - : Indicates how the violated policy is configured to be treated by the user agent. This will be `"enforce"` or `"report"`.
- {{domxref("CSPViolationReportBody.documentURL")}} {{ReadOnlyInline}} {{Experimental_Inline}}
  - : A string representing the URL of the document or worker in which the violation was found.
- {{domxref("CSPViolationReportBody.effectiveDirective")}} {{ReadOnlyInline}} {{Experimental_Inline}}
  - : A string representing the directive whose enforcement uncovered the violation.
- {{domxref("CSPViolationReportBody.lineNumber")}} {{ReadOnlyInline}} {{Experimental_Inline}}
  - : The line number in the document or worker at which the violation occurred.
- {{domxref("CSPViolationReportBody.originalPolicy")}} {{ReadOnlyInline}} {{Experimental_Inline}}
  - : A string containing the policy whose enforcement uncovered the violation.
- {{domxref("CSPViolationReportBody.referrer")}} {{ReadOnlyInline}} {{Experimental_Inline}}

>   - : A string representing the referrer of the resources whose policy was violated. This will be a URL or `null`.

Reviewer 2: Pedantically, it's not a string if it's `null`. SO maybe:

```suggestion
  - : A string representing the the URL for referrer of the resources whose policy was violated, or `null`.
```

- {{domxref("CSPViolationReportBody.sample")}} {{ReadOnlyInline}} {{Experimental_Inline}}
  - : A string representing a sample of the resource that caused the violation, usually the first 40 characters. This will only be populated if the resource is an inline script, event handler, or style — external resources causing a violation will not generate a sample.
- {{domxref("CSPViolationReportBody.sourceFile")}} {{ReadOnlyInline}} {{Experimental_Inline}}

>   - : A string representing the URL of the document or worker in which the violation was found.

Author: These should probably be updated in `SecurityPolicyViolationEvent` as well; I had copied most of these descriptions from there since the two interfaces are so similar.
